working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+# Primary Constructors
+
+Author: Erik Ernst
+
+Status: Draft
+
+Version: 1.0
+
+Experiment flag: primary-constructors
+
+This document specifies _primary constructors_. This is a feature that allows
+one constructor and a set of instance variables to be specified in a concise
+form in the header of the declaration. In order to use this feature, the given
+constructor must satisfy certain constraints, e.g., it cannot have a body.
+
+One variant of this feature has been proposed in the [struct proposal][],
+several other proposals have appeared elsewhere, and prior art exists in
+languages like [Kotlin][kotlin primary constructors] and Scala (with
+specification [here][scala primary constructors] and some examples
+[here][scala primary constructor examples]). Many discussions about the
+feature have taken place in github issues marked with the
+[primary-constructors label][].
+
+[struct proposal]: https://github.com/dart-lang/language/blob/master/working/extension_structs/overview.md
+[kotlin primary constructors]: https://kotlinlang.org/docs/classes.html#constructors
+[scala primary constructors]: https://www.scala-lang.org/files/archive/spec/2.11/05-classes-and-objects.html#constructor-definitions
+[scala primary constructor examples]: https://www.geeksforgeeks.org/scala-primary-constructor/
+[primary-constructors label]: https://github.com/dart-lang/language/issues?q=is%3Aissue+is%3Aopen+primary+constructor+label%3Aprimary-constructors
+
+## Introduction
+
+Primary constructors is a conciseness feature. It does not provide any new
+semantics at all. It just allows us to express something which is already
+possible in Dart, using a less verbose notation. Consider this sample class
+with two fields and a constructor:
+
+```dart
+// Current syntax.
+
+class Point {
+  int x;
+  int y;
+  Point(this.x, this.y);
+}
+```
+
+A primary constructor allows us to define the same class much more
+concisely:
+
+```dart
+// A declaration with the same meaning, using a primary constructor.
+
+class Point(int x, int y);
+```
+
+In the examples below we show the current syntax directly followed by a
+declaration using a primary constructor. The meaning of the two class
+declarations with the same name is always the same. Of course, we would
+have a name clash if we actually put those two declarations into the same
+library, so we should read the examples as "you can write this _or_ you can
+write that". So the example above would be shown as follows:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, this.y);
+}
+
+class Point(int x, int y);
+```
+
+These examples will serve as an illustration of the proposed syntax, but
+they will also illustrate the semantics of the primary constructor
+declarations, because those declarations work exactly the same as the
+declarations using the current syntax.
+
+Note that an empty class body, `{}`, can be replaced by `;`.
```

Comment:
ACK. I'm considering whether we want to remove `class C = S with M;` from the language, and perhaps let `class C extends S with M;` to be the new syntax for the same thing. 
That obviously only works when we are not introducing members into `C` with a primary constructor.

The distinction between `class C extends Object with M {}` and `class C = Object with M;` is that the latter can be a mixin class, it has `Object` as superclass, with `M` mixed in on top, the former cannot because it has `Object&M` as superclass.

(But we could also just expand the `mixin` declaration to allow defining composite mixins, then renaming a mixin and adding interfaces is just the trivial case of a singleton composite mixin.)

Reply:
> considering whether we want to remove `class C = S with M;` from the language [and replacing it]

Maybe we could simply remove that syntax. The fact that it is no longer possible to create exactly the same entity may not be a problem, and developers could use `class C extends S with M;` as a very-very-very-nearly equivalent replacement.

> composite mixins

That would be great!

Reply:
In any case, it sounds like there is no big conflict on letting `class C extends S with M;` mean `class C extends S with M {}`.